---
title: "NMT-Seq: Single-cell accessibility profiles with confidence intervals"
output: 
  BiocStyle::html_document: 
    fig_width: 12
    fig_height: 8
---

```{r echo=FALSE, include=FALSE}
# devtools::install_github("andreaskapou/BPRMeth-devel")
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(BPRMeth))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(stringi))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(gridExtra))
```

```{r echo=FALSE, include=FALSE}
ggplot_bpr_profiles <- function(X, obs, title="", subtitle="", up_label="-500bp", down_label="+500bp", middle_label="TSS", ...){
  # ggplot(X, aes(x = xs, y = ys)) +
  p <- ggplot(data = data.frame(X), aes(x = xs, y = ys)) +
    geom_line(aes(x = xs, y = ys), size=1.5, col = "darkblue") +
    geom_point(data = obs, mapping = aes(x = x, y = y), shape=1, color="red", size=3) +
    geom_ribbon(data = data.frame(X), mapping = aes(ymin= ys_low, ymax= ys_high), alpha=0.4, size = 0.1, fill="cornflowerblue") +
    scale_x_continuous(limits = c(-1, 1), labels =c(up_label, "", middle_label, "", down_label)) + 
    scale_y_continuous(limits = c(0, 1), breaks=pretty_breaks(n=4)) + 
    labs(title = title, subtitle = subtitle, x="genomic region", y="Accessibility") + 
    line_theme()# + theme(axis.text.x = element_blank())
}

# Define ggplot2 theme for line plots
line_theme <- function(){
  p <- theme(
      plot.title=element_text(size=20, face='bold', margin=margin(0,0,3,0), hjust=0.5),
      #plot.subtitle=element_text(size=15, margin=margin(0,0,4,0), hjust=0.5),
      axis.text=element_text(size=rel(1.05), color='black'),
      axis.title=element_text(size=rel(1.45), color='black'),
      axis.title.y = element_text(margin=margin(0,10,0,0)),
      axis.title.x = element_text(margin=margin(10,0,0,0)),
      axis.ticks.x = element_line(colour="black", size=rel(0.8)),
      axis.ticks.y = element_blank(),
      legend.position="right",
      legend.key.size = unit(1.9, 'lines'),
      legend.title=element_text(size=24, face='bold'),
      legend.text=element_text(size=19),
      panel.border=element_blank(),
      panel.grid.major = element_line(colour = "gainsboro"),
      #panel.grid.minor = element_line(colour = "grey"),
      panel.background = element_blank()
    )
}
```



<!-- # Parse and filter data -->
```{r echo=FALSE, include=FALSE}
# Data
set.seed(12345)
io                     <- list()
io$base_dir            <- "/home/cakapourani/Documents/Code/datasets/nmt-seq-data-eb"
load(paste0(io$base_dir,"/acc/parsed/profiles/acc_400bp.RData"))
io$base_dir            <- "/home/cakapourani/Documents/Code/datasets/nmt-seq-data-eb"
io$out_dir             <- "/home/cakapourani/Documents/Code/nmt-seq/rebuttal/EB/acc_conservation/out/examples/"
io$rna_file            <- paste0(io$base_dir, "/rna/parsed/sceset.rds")
opts                   <- list(basis_prof = create_rbf_object(M = 13), gpc_cov = 5)
```

<!-- # Load scRNA data -->
```{r load_rna, echo=FALSE, include=FALSE, warning=FALSE, message=FALSE}
sceset       <- readRDS(file = io$rna_file)  # Load expression as SCESet object
# Create data.table
rna_dt       <- exprs(sceset) %>% t %>% as.data.table(keep.rownames = "sample") %>% melt(id.vars = "sample", value.name = "expr", variable.name = "gene")
# Compute summary statistics for each gene across cells
rna_all_st   <- rna_dt[, .(mean_expr = mean(expr), median_expr = median(expr), var_expr = var(expr), sd_expr = sd(expr), cv_expr = sd(expr)/mean(expr)), by = gene]
N_cells      <- length(unique(rna_dt$sample))  # Number of unique cells
rna_all_st   <- rna_all_st[, cv_expr := log2((1 + 1/(4 * N_cells)) * cv_expr)]  # Correct CV empirically

# Extract gene coordinates metadata to do the overlap
rna_metadata <- fData(sceset) %>% tibble::rownames_to_column("gene") %>% as.data.table %>% .[,c("chr", "start", "end", "gene", "ens_id")]  %>% .[, chr := as.factor(sub("chr", "", chr))] %>% setnames("ens_id", "id")

rna_all_st   <- merge(rna_metadata[, c("chr", "id", "gene")], rna_all_st, by = c("gene")) # Merge using all samples
rna_dt       <- merge(rna_metadata[, c("chr", "id", "gene")], rna_dt, by = c("gene"))     # Merge using only samples that pass filtering
rm(sceset, rna_metadata)
```

# Learn GC accessibility profiles using BPRMeth model
```{r dnarna_bprfit, echo=FALSE, include=TRUE, warning=FALSE, fig.align='center', fig.width=13, fig.height=7}
# Keep regions with enough GpC coverage
region_dt <- lapply(region_dt, function(x) lapply(x, function(y){ if(NROW(y)<opts$gpc_cov) return(NA) else return(y) }))

# id <- "ENSMUSG00000000730"
gene_id <- "ENSMUSG00000036120"
gene_id <- "ENSMUSG00000003575"
gene_id <- "ENSMUSG00000003198"
gene_id <- "ENSMUSG00000000303"
gene_id <- "ENSMUSG00000006273"

gene_id <- "ENSMUSG00000021255"
# Get specific gene
gene <- lapply(region_dt, "[[", gene_id)
gene <- gene[-which(is.na(gene))]
cells_names <- names(gene)
# Concatenate to obtain bulk data
bulk <- do.call(rbind, gene)

# Learn profiles for each cell
prof_cells <- bpr_bayes(x = gene, basis = opts$basis_prof, gibbs_nsim = 3000, gibbs_burn_in = 1000, is_parallel = FALSE, keep_gibbs_draws = TRUE)
# Learn bulk profile
prof_bulk <- bpr_bayes(x = bulk, basis = opts$basis_prof, gibbs_nsim = 3000, gibbs_burn_in = 1000, keep_gibbs_draws = TRUE)
# Concatenate all results to one object
prof_cells$W_opt <- rbind(prof_cells$W_opt, prof_bulk$w_opt)
prof_cells$W_var <- rbind(prof_cells$W_var, prof_bulk$w_var)
prof_cells$W_draws[[length(prof_cells$W_draws) + 1]] <- prof_bulk$w_draws

# Initialize
xs = seq(-1, 1, length = 200)
H <- design_matrix(x = opts$basis_prof, obs = xs)$H
pp <- list()
ys_draws <- matrix(0, nrow = NROW(prof_bulk$w_draws), ncol = length(xs))
dt <- data.table(xs = xs, ys = 0, ys_low = 0, ys_high = 0)

# Create subdirectories for each gene
io$out_dir_loc <- paste0(io$out_dir, gene_id, "/")
ifelse(!dir.exists(file.path(io$out_dir_loc)), dir.create(file.path(io$out_dir_loc)), FALSE)


# pdf(file = paste0(out_dir, "acc_profiles_", gene_id,  ".pdf"), onefile = TRUE, width = 12, height = 7)
for (m in 1:length(prof_cells$W_draws)){
  w_draws <- prof_cells$W_draws[[m]] # Compute predictive distribution
  for (i in 1:NROW(prof_cells$W_draws[[1]])){ ys_draws[i, ] <- pnorm(H %*% w_draws[i, ]) }
  # Compute quantiles of ys
  ys_q <- apply(ys_draws, 2, quantile, probs = c(0.1, 0.9),  na.rm = TRUE)
  dt <- dt %>% .[, c("ys", "ys_low", "ys_high") := list(colMeans(ys_draws), ys_q[1, ], ys_q[2, ])]
  obs <- as.data.table(gene[[m]]) %>% setnames(c("x", "y"))
  pp[[m]] <- ggplot_bpr_profiles(X = dt, obs = obs, title = paste0("Cell ", cells_names[m]),  up_label = "-200bp", down_label = "+200bp", middle_label = "TSS") ##subtitle = paste0("Expr = ", round(rna_dt[id == gene_id & sample == cells_names[m], expr], 3))
  
  pdf(file = paste0(io$out_dir_loc, "cell_", cells_names[m], "_prof_", gene_id, ".pdf"), width = 6, height = 4)
  print(pp[[m]])
  dev.off()
}
```
